### PR TITLE
feat: create bound container from text

### PR DESCRIPTION
**Pull request by ad1992**
closes https://github.com/excalidraw/excalidraw/issues/5670

https://user-images.githubusercontent.com/11256141/222403703-6bc4f5b2-826f-43ce-a683-41823e7f43c8.mp4

- [x] Add tests
- [x] There was a corner case where when splitting words, and if they add up to exact `max width` it was still splitting it, that's fixed now.
eg max width of container = 100
The text is "Hello world there"
Width of world + there = 100 still it will split it was splitting it to `Hello\nWorld\nthere`
- [x] consider groups
- [x] consider bindings

### CODE REVIEW

File: src/actions/actionBoundText.tsx [hunk 1]
```diff
@@ -129,18 +138,105 @@ export const actionBindText = register({
...
+  const textElementIndex = updatedElements.findIndex(
+    (ele) => ele.id === textElement.id,
+  );
+  updatedElements.splice(textElementIndex, 1);
+  const containerIndex = updatedElements.findIndex(
+    (ele) => ele.id === containerId,
+  );
+  updatedElements.splice(containerIndex + 1, 0, textElement);
+  return updatedElements;
+};
```
Comment (dwelle) [inline src/actions/actionBoundText.tsx:165]: in case of creating a container from text, it should put the container below the text index, not the other way round as with binding text to an existing container action.

Comment (ad1992) [reply src/actions/actionBoundText.tsx:165]: How does it make a difference as I am pushing the text element above the container? I am reusing the function for both cases

**Comment by dwelle** [comment]
Since the goal seems to be not to break the text by wrapping into a smaller container than text's width, we should account for padding of the container so it doesn't break at all:

excal-create-container

File: src/locales/en.json [hunk 2]
```diff
@@ -110,6 +110,7 @@
     "increaseFontSize": "Increase font size",
     "unbindText": "Unbind text",
     "bindText": "Bind text to the container",
+    "createContainerFromText": "Create Container from text",
```
Comment (dwelle) [inline src/locales/en.json:113]:
```suggestion
    "createContainerFromText": "Wrap text in a container",
```

File: src/element/textElement.ts [hunk 3]
```diff
@@ -714,34 +710,53 @@ export const getTextBindableContainerAtPosition = (
...
+  }
+  if (containerType === "arrow") {
+    return boundTextElementWidth + BOUND_TEXT_PADDING * 8 * 2;
+  }
+  if (containerType === "diamond") {
+    return 2 * (boundTextElementWidth + BOUND_TEXT_PADDING * 2);
+  }
+  return boundTextElementWidth + BOUND_TEXT_PADDING * 2;
+};
+
```
Comment (rahul3v) [inline src/element/textElement.ts:759]: why not use it as `computeContainerForBoundText` and pass height/width as needed, 
as both functions are functionally doing the same

File: src/element/textElement.ts [hunk 4]
```diff
@@ -714,34 +710,53 @@ export const getTextBindableContainerAtPosition = (
...
+const VALID_CONTAINER_TYPES = [
+  "rectangle",
+  "ellipse",
+  "diamond",
+  "image",
+  "arrow",
+];
+
+export const isValidTextContainer = (element: ExcalidrawElement) =>
+  VALID_CONTAINER_TYPES.includes(element.type);
```
Comment (rahul3v) [inline src/element/textElement.ts:722]:
```suggestion
const VALID_CONTAINER_TYPES = new Set([
  "rectangle",
  "ellipse",
  "diamond",
  "image",
  "arrow",
]);

export const isValidTextContainer = (element: ExcalidrawElement) =>
  VALID_CONTAINER_TYPES.has(element.type);
```

Set has a time complexity o(1)

**Comment by ad1992** [comment]
> Since the goal seems to be not to break the text by wrapping into a smaller container than text's width, we should account for padding of the container so it doesn't break at all:
> 
> excal-create-container [ excal-create-container ](https://user-images.githubusercontent.com/5153846/222418306-574fea2f-90fe-4b2c-baa3-da8321615f27.gif) [ ](https://user-images.githubusercontent.com/5153846/222418306-574fea2f-90fe-4b2c-baa3-da8321615f27.gif)

Though I was considering padding, this issue was due to the dimensions being exact decimals so now I am rounding it off and should work now, can you try it out and confirm ?

File: src/element/textElement.ts (hunk 3 above)
Comment (rahul3v) [reply src/element/textElement.ts]: also you can short things out `boundTextElementWidth/Height + BOUND_TEXT_PADDING * 2` is **common**, you can pre-calculate it before and use it down eleswhere

Comment (rahul3v) [reply src/element/textElement.ts]: you can short the code little more `boundTextElementWidth/Height + BOUND_TEXT_PADDING * 2` is mostly **common**, you can per-calculate and use it down elsewhere

File: src/actions/actionBoundText.tsx (hunk 1 above)
Comment (dwelle) [reply src/actions/actionBoundText.tsx:164]: we should respect the original z-index (of the text in this case):

excal-wrap-into-container-z-index-issue

File: src/element/textElement.ts [hunk 5]
```diff
@@ -714,32 +709,36 @@ export const getTextBindableContainerAtPosition = (
...
+  "ellipse",
+  "diamond",
+  "image",
+  "arrow",
+]);
 
-export const computeContainerHeightForBoundText = (
-  container: NonDeletedExcalidrawElement,
-  boundTextElementHeight: number,
+export type ValidContainerType<T> = T extends Set<infer U> ? U : never;
```
Comment (dwelle) [inline src/element/textElement.ts:720]: let's move this to `src/global.d.ts` and make into a utility type. And could be slightly improved:

```ts
type ExtractSetType<T extends Set<any>> = T extends Set<infer U> ? U : never;
```

File: src/actions/actionBoundText.tsx (hunk 1 above)
Comment (ad1992) [reply src/actions/actionBoundText.tsx]: True, I have fixed it

**Comment by dwelle** [comment]
- [ ] should account for groups:

    excal-wrap-into-container-group-issue

**Review by dwelle — APPROVED** [review]
